src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -993,6 +993,11 @@ private static SizeAndAlignment ComputeInstanceSize(MetadataType type, LayoutInt
                 }
             }
 
+            if (type.ContainsGCPointers)
+            {
+                instanceSize = LayoutInt.AlignUp(instanceSize, new LayoutInt(target.PointerSize), target);
```

Comment:
Does the general code here also need factor in other higher packing like for `Vector64/128/256/512`, `Int128/UInt128`, or `long/ulong/double`?

It seems odd in general how setting `Size` works (below is for `Sequential`, but `Explicit` is generally the same).

Setting `Size = 0` is ignored:
```csharp
[StructLayout(LayoutKind.Sequential, Size = 0)]
struct MyStruct      // Size 16
{
    public double a; // offset: 0
    public int b;    // offset: 8
    public byte c;   // offset: 12
}
```

Setting `Size < NaturalSize` trims any packing, but doesn't actually change the layout/packing, but trims the size:
```csharp
[StructLayout(LayoutKind.Sequential, Size = 1)]
struct MyStruct      // Size 13
{
    public double a; // offset: 0
    public int b;    // offset: 8
    public byte c;   // offset: 12
}
```

Setting `Size >= NaturalSize` just adds more bytes, but similarly doesn't make things line up for packing.

Reply:
@tannergooding I presume you tested on CoreCLR, right? I guess for those case we need to file a separate issue - this PR fixes an obvious bug that is only NAOT/R2R specific

Reply:
> I presume you tested on CoreCLR, right

Yes

> his PR fixes an obvious bug that is only NAOT/R2R specific

I don't think it's NAOT/R2R specific. A new console app containing the repro from #100220 prints the same thing for RyuJIT (`9`, when it should be `16`).